### PR TITLE
Disable NVJPEG HW decoder for driver < 455 due to performance reason

### PR DESCRIPTION
**Pull request by JanuszL**
- Due to frequent memory allocations NVJPEG HW decoder on driver < 455,
  where memory pool is not available in the driver, hits the overall
  performance. Due to this disable it.

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [x] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

- Due to frequent memory allocations NVJPEG HW decoder on driver < 455, where memory pool is not available in the driver, hits the overall performance. Due to this disable it.

#### Additional information
- Affected modules and functionalities:
  nvjpeg_decoder_decoupled_api

- Key points relevant for the review:
  NA

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h [hunk 1]
```diff
@@ -97,6 +98,25 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
 
     if (try_init_hw_decoder &&
         nvjpegCreate(NVJPEG_BACKEND_HARDWARE, NULL, &handle_) == NVJPEG_STATUS_SUCCESS) {
+    // disable HW decoder for drivers < 455.x as the memory pool for it is not available
+    // and multi GPU performance is far from perfect due to frequent memory allocations
+#if NVML_ENABLED
+    char version[80];
```
Comment (klecki) [inline dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h:104]: Can you access the `nvmlConstants::NVML_SYSTEM_DRIVER_VERSION_BUFFER_SIZE`? (I know the docs says it's 80, I don't know which one is better).

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h [hunk 2]
```diff
@@ -97,6 +98,25 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
 
     if (try_init_hw_decoder &&
         nvjpegCreate(NVJPEG_BACKEND_HARDWARE, NULL, &handle_) == NVJPEG_STATUS_SUCCESS) {
+    // disable HW decoder for drivers < 455.x as the memory pool for it is not available
```
Comment (klecki) [inline dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h:101]: I wonder if it makes sense to try to initialize HW decoder and than switch to regular NVJPEG handle as you still go through all of the config.

Isn't it better to do it in the check above - if we have hw_decoder_load specified, but the driver is old, just set the load to 0 and try_init_hw_decoder to false and don't bother with the setup here.

Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h:101]: I wanted to show a warning that there is a HW decoder but it is disabled. The best way is to just call `nvjpegCreate(NVJPEG_BACKEND_HARDWARE, NULL, &handle_)` to make sure the HW decoder is available.

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h (hunk 1 above)
Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h:104]: Done

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h (hunk 2 above)
Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h:101]: Anyway, I forgot to skip the rest of the code if the driver is not sufficient.

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by klecki — APPROVED** [review]

**Review by szalpal — APPROVED** [review]

**Pull request merged**